File: c++/c15_1_kruskals_algorithm.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "util.h"
#include "c12_3_applications.h"
#include "c15_2_union_find_structure.h"

namespace {
	vector<vector<pair<ll, ll>>> kruskal(const vector<vector<pair<ll, ll>>>& adj) {
		ll n = ssize(adj);
		vector<tuple<ll, ll, ll>> edges;
		for (ll a = 1; a < n; ++a)
			for (auto [b, w] : adj[a])
				edges.emplace_back(w, a, b);
		ranges::sort(edges);
		UnionFind uf(n);
		vector<vector<pair<ll, ll>>> ret(n);
		for (auto [w, a, b] : edges) {
			if (!uf.same(a, b)) {
				uf.unite(a, b);
				ret[a].emplace_back(b, w);
				ret[b].emplace_back(a, w);
			}
		}
		return ret;
	}
}

TEST(C151KruskalsAlgorithm, kruskal) {
	EXPECT_EQ(kruskal({ {} }), (vector<vector<pair<ll, ll>>>{ {} }));
	auto ret = kruskal({ {},{{2,3},{5,5}},{{1,3},{3,5},{5,6}},{{2,5},{4,9},{6,3}},{{3,9},{6,7}},{{1,5},{2,6},{6,2}},{{3,3},{4,7},{5,2}} });
	ll sum = 0;
	for (const auto& v : ret)
		for (auto [_, w] : v)
			sum += w;
	EXPECT_EQ(sum, 20 * 2);
	auto adj = delweight(ret);
	EXPECT_EQ(connected(adj).size(), 1);
	EXPECT_FALSE(finding_cycles(adj));
}

File: python/c15_1_kruskals_algorithm.py
import unittest
from util import delweight
from c12_3_applications import connected,finding_cycles
from c15_2_union_find_structure import UnionFind

def kruskal(adj):
    n=len(adj)
    ret=[[] for _ in adj]
    uf=UnionFind(n)
    for w,a,b in sorted((w,a,b) for a in range(1,n) for b,w in adj[a]):
        if not uf.same(a,b):
            uf.unite(a,b)
            ret[a].append((b,w))
            ret[b].append((a,w))
    return ret

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(kruskal([[]]),[[]])
        ret=kruskal([
            [],
            [(2,3),(5,5)],
            [(1,3),(3,5),(5,6)],
            [(2,5),(4,9),(6,3)],
            [(3,9),(6,7)],
            [(1,5),(2,6),(6,2)],
            [(3,3),(4,7),(5,2)]
        ])
        self.assertEqual(sum(w for i in range(1,len(ret)) for _,w in ret[i]),20*2)
        adj=delweight(ret)
        self.assertEqual(len([*connected(adj)]),1)
        self.assertFalse(finding_cycles(adj))

if __name__=='__main__':
    unittest.main()
